tests/Microsoft.Graph.Core.Test/Requests/BaseRequestTests.cs

```diff
@@ -165,6 +166,50 @@ public void GetWebRequest_OverrideCustomTelemetryHeader()
             }
         }
 
+        [TestMethod]
+        public async Task SendAsync_ResponseHeaders()
+        {
+            var requestUrl = string.Concat(this.baseUrl, "/me/drive/items/id");
+
+            var baseRequest = new BaseRequest(requestUrl, this.baseClient) { ContentType = "application/json" };
+
+            var data = "{\"data\"}";
+            
+            using (var httpResponseMessage = new HttpResponseMessage())
+            using (var responseStream = new MemoryStream(Encoding.ASCII.GetBytes(data)))
+            using (var streamContent = new StreamContent(responseStream))
+            {
+                httpResponseMessage.Content = streamContent;
+                httpResponseMessage.StatusCode = System.Net.HttpStatusCode.OK;
+
+                this.httpProvider.Setup(
+                    provider => provider.SendAsync(
+                        It.Is<HttpRequestMessage>(
+                            request =>
+                                string.Equals(request.Content.Headers.ContentType.ToString(), "application/json")
+                               && request.RequestUri.ToString().Equals(requestUrl)),
+                        HttpCompletionOption.ResponseContentRead,
+                        CancellationToken.None))
+                        .Returns(Task.FromResult(httpResponseMessage));
+
+                Dictionary<string, object> additionalData = new Dictionary<string,object>();
+                additionalData["responseHeaders"] = new Dictionary<string, List<string>>() { { "key", new List<string>() { "value" } } };
+
+                var expectedResponseItem = new DerivedTypeClass { Id = "id" , AdditionalData = additionalData };
+
+                this.serializer.Setup(
+                    serializer => serializer.DeserializeObject<DerivedTypeClass>(It.IsAny<string>()))
+                    .Returns(expectedResponseItem);
+                this.serializer.Setup(
+                    serializer => serializer.DeserializeObject<DerivedTypeClass>(It.IsAny<string>()))
+                    .Returns(expectedResponseItem);
+
+                var responseItem = await baseRequest.SendAsync<DerivedTypeClass>("string", CancellationToken.None);
+                Assert.IsNotNull(responseItem.AdditionalData["responseHeaders"], "No response headers available");
+                Assert.AreEqual(expectedResponseItem.AdditionalData["responseHeaders"], responseItem.AdditionalData["responseHeaders"], "Unexpected response headers");
+            }
```

Comment:
Looks good as far as I see. Can you walk me through this so that I can understand how the mock library works.